### PR TITLE
Add SourceContainerNameFrom and EnvNames in sidecarset transferenv

### PR DESCRIPTION
**Pull request by pigletfly**
Signed-off-by: pigletfly <wangbing.adam@gmail.com>

### Ⅰ. Describe what this PR does

as title said

### Ⅱ. Does this pull request fix one issue?

fixes of https://github.com/openkruise/kruise/issues/818

### CODE REVIEW

File: pkg/control/sidecarcontrol/util.go [hunk 1]
```diff
@@ -270,13 +277,136 @@ func GetSidecarTransferEnvs(sidecarContainer *appsv1alpha1.SidecarContainer, pod
...
+	if err != nil {
+		return "", err
 	}
+
+	accessor, err := meta.Accessor(pod)
+	if err != nil {
+		return "", err
+	}
+
+	if path, subscript, ok := SplitMaybeSubscriptedPath(fieldPath); ok {
```
[the remark this answers is not in the document]
Comment (pigletfly) [reply pkg/control/sidecarcontrol/util.go:317]: fixed. code from lifted.

[the remark this answers is not in the document]
Comment (FillZpp) [reply pkg/control/sidecarcontrol/util.go:317]: Still don't understand why need to copy and re-define a `SplitMaybeSubscriptedPath` function here? Kubernetes dependency is already in project, we can directly use `fieldpath.SplitMaybeSubscriptedPath`.

[the remark this answers is not in the document]
Comment (pigletfly) [reply pkg/control/sidecarcontrol/util.go]: ok.My intention is avoid direct import k/k.

**Comment by zmberg** [comment]
@pigletfly please add same UT and e2e case

**Comment by pigletfly** [comment]
> 

ok, will do it ASAP

**Comment by pigletfly** [comment]
@zmberg @FillZpp PTAL

**Comment by FillZpp** [comment]
LGTM, waiting for the tests.

**Comment by FillZpp** [comment]
@pigletfly seems the new e2e case has failed

**Comment by pigletfly** [comment]
All green now

File: pkg/control/sidecarcontrol/util.go [hunk 2]
```diff
@@ -270,13 +274,101 @@ func GetSidecarTransferEnvs(sidecarContainer *appsv1alpha1.SidecarContainer, pod
...
+			}
+			injectedEnvs = append(injectedEnvs, env)
 		}
-		injectedEnvs = append(injectedEnvs, env)
 	}
 	return
 }
+
+func ExtractContainerNameFromFieldPath(fs *corev1.ObjectFieldSelector, pod *corev1.Pod) (string, error) {
+	fieldPath, _, err := ConvertDownwardAPIFieldLabel(fs.APIVersion, fs.FieldPath, "")
```
Comment (zmberg) [inline pkg/control/sidecarcontrol/util.go:309]: Can this method refer directly to the K8S native method?

Comment (pigletfly) [reply pkg/control/sidecarcontrol/util.go:309]: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/pods/helpers.go#L79-L93 the origin function supports more than we needed(metadata.annotations,metadata.labels,metadata.name)

Comment (zmberg) [reply pkg/control/sidecarcontrol/util.go:309]: Lines 336-344 in the uitl.go file have been verified. Make sure we only support these fields, ok?

Comment (pigletfly) [reply pkg/control/sidecarcontrol/util.go:309]: yes,I think container name won't appear in pod status.

**Comment by zmberg** [comment]
good.
/lgtm

File: pkg/control/sidecarcontrol/util.go [hunk 3]
```diff
@@ -270,13 +274,80 @@ func GetSidecarTransferEnvs(sidecarContainer *appsv1alpha1.SidecarContainer, pod
...
+		sourceContainerName := tEnv.SourceContainerName
+		if tEnv.SourceContainerNameFrom != nil && tEnv.SourceContainerNameFrom.FieldRef != nil {
+			containerName, err := ExtractContainerNameFromFieldPath(tEnv.SourceContainerNameFrom.FieldRef, pod)
+			if err != nil {
+				klog.Errorf("unmarshal pod(%s.%s) annotations[%s] failed: %s", pod.Namespace, pod.Name, err.Error())
+				continue
+			}
+			sourceContainerName = containerName
+		}
+		// if sourceContainerName is empty or not found in pod.spec.containers
```
Comment (furykerry) [inline pkg/control/sidecarcontrol/util.go:294]: "if sourceContainerName is empty or not found in pod.spec.containers" is this comments valid here, consider move above the warning log

File: pkg/control/sidecarcontrol/util.go [hunk 4]
```diff
@@ -270,13 +274,80 @@ func GetSidecarTransferEnvs(sidecarContainer *appsv1alpha1.SidecarContainer, pod
...
+	if version != "v1" {
+		return "", "", fmt.Errorf("unsupported pod version: %s", version)
+	}
+	path, _, ok := fieldpath.SplitMaybeSubscriptedPath(label)
+	if ok {
+		switch path {
+		case "metadata.annotations", "metadata.labels":
+			return label, value, nil
+		default:
+			return "", "", fmt.Errorf("field label does not support subscript: %s", label)
```
Comment (furykerry) [inline pkg/control/sidecarcontrol/util.go:349]: the error message seems a bit confusing, consider replace the  error "field path not supported: %s", path

**Pull request merged**